src/libraries/System.Collections.Immutable/ref/System.Collections.Immutable.cs

```diff
@@ -223,24 +344,33 @@ public sealed partial class Builder : System.Collections.Generic.ICollection<T>,
             public bool Contains(T item) { throw null; }
             public void CopyTo(T[] array, int index) { }
             public System.Collections.Generic.IEnumerator<T> GetEnumerator() { throw null; }
+            [System.Diagnostics.Contracts.PureAttribute]
```

Comment:
Is it by design that Pure isn't on members like Contains, CopyTo, etc.?

Reply:
Good point.  I can confirm that we're correctly propogating what was in the source, but it does seem inconsistent that the source only applies this attribute to some methods.  Contains and CopyTo should be Pure, whereas I think there are some other methods marked `Pure` that should not be.

Reply:
I suspect Contains was supposed to inherit its Pure attribute from the interface that defined it.
CopyTo, on the other hand, is _not_ Pure. It mutates existing objects. That's the whole point.

Reply:
> CopyTo, on the other hand, is not Pure. It mutates existing objects. That's the whole point.

Yet it's on ImmutableArray.CopyTo?

Reply:
There's a lot of inconsistentcy here. ImmutableDictionary.Contains, ContainsKey are both lacking it. ID.Clear appears to be following the immutable pattern but it's not marked Pure. ImmutableList.BinarySearch isn't Pure?

I don't like that we're promoting a dead feature set here. But if we're going to do it we should do it right.

Reply:
Why are we saying it's dead? `PureAttribute` is just an attribute, for which analyzers (and/or compiler warnings) exist to help people avoid mistakes that are quite common when dealing with immutable collections. The rest of the namespace is dead sure, like `Contracts.Requires`. But IMO that shouldn't undermine the value of PureAttribute.
As for doing it right, sure: let's fix the missing or inappropriate attributes.

@ericstj how do you want to handle this? Do you want to push fixes for these to your PR? Would you like *me* to audit and push to your PR? Push to a separate PR?

Reply:
> Why are we saying it's dead?

- It’s a sliver of a Contracts system/namespace that was never fully productized and has long fallen by the wayside.  The primary purpose of PureAttribute was to mark methods that could be safely used in other Contract usage. And it requires you to set CONTRACTS_FULL to even show up in the compiled library, which almost no one does even if they’re using only Pure, and which will cause the rest of the Contracts library to fail without the binary rewriter, again highlighting the purpose of the attribute as being just for contracts.
- There is no validation that what it’s applied to is actually pure, as evidenced by the multitude of incorrect usage in this library.
- There’s tons of pure methods throughout both the core libraries and external libraries, and its used on only a very, very tiny portion of them, including in dotnet/runtime, where out of the hundreds of libraries present, it’s basically only used by System.Collections.Immutable.
- It’s ambiguous in its application and lacks the ability to appropriately express things correctly.  For example, S.C.Immutable applies it to its LINQ methods like Any, All, Aggregate, etc. that take Func<…> predicates and invoke those predicates… unless the caller has also guaranteed that those funcs are pure, these calls are not.
- The only practical use I’ve seen for it is for an analyzer that flags “you should use this method’s return value”, but there are plenty of methods that aren’t pure where ignoring the return value is almost certainly a mistake, too, e.g. Stream.Read. Even Midori, which has a robust “permissions” system, much more extensive than whether something was pure or not, and which prohibited global state mutations by construction, still recognized the ill-advised nature of combining these concepts and had a separate [DoNotIgnore] attribute that could be applied to return values, out parameters, etc.

Reply:
hmmm... You make some very good points. Do you see more negative value in this PR than positive?

Reply:
Personally, I'd prefer to see us not do this, and instead remove [Pure] from the implementation.  Then if these are scenarios we believe are important, take a step back and design attributes and associated analyzers (or language features) we believe are important here.  If it turns out [Pure] is in fact what we want, then we can re-evaluate, but I suspect it's not.

Reply:
I can live with that.

Reply:
I’ll update this PR to remove Pure then.